jib-core/src/main/java/com/google/cloud/tools/jib/image/ReproducibleLayerBuilder.java

```diff
@@ -100,9 +100,11 @@ public Blob build() {
     for (LayerEntry layerEntry : layerEntries) {
       // Adds the entries to uniqueTarArchiveEntries, which makes sure all entries are unique and
       // adds parent directories for each extraction path.
-      uniqueTarArchiveEntries.add(
+      TarArchiveEntry entry =
           new TarArchiveEntry(
-              layerEntry.getSourceFile().toFile(), layerEntry.getAbsoluteExtractionPathString()));
+              layerEntry.getSourceFile().toFile(), layerEntry.getAbsoluteExtractionPathString());
+      entry.setMode((entry.getMode() & ~0777) | layerEntry.getPermissions());
```

Comment:
`entry.getMode() & ~0777`

This looks weird. I think you are always resetting the right-most three bits to 0, effectively making the number to 0.

Also, are you sure `TarArchiveEntry.setMode(int)` expects an octal number? The Javadoc doesn't say anything about it, but I think it's very possible that `setMode()` just expects a decimal number.

Reply:
I'll change it for clarity, but to clear things up:

>are you sure TarArchiveEntry.setMode(int) expects an octal number

Numbers starting with 0 are treated as octal in java; i.e. `0777` and `511` are treated as the same thing. So you're right that `TarArchiveEntry.setMode` expects a decimal number, but passing in an octal representation doesn't make a difference as long as it's specified as octal.

>This looks weird. I think you are always resetting the right-most three digits to 0, effectively making the number to 0.

The reason for the inversion in `~0777` is to make the lower 9 bits `0` and the remaining upper bits `1` so only the permission bits are masked out, because the `mode` int contains more than just permissions.

Reply:
Okay, let me rephrase my question: so, in order to set the executable, write, and read bits to everything (i.e., the usual "777" numeric notation), should it be `TarArchiveEntry.setMode(777)` or `TarArchiveEntry.setMode(511)`? I thought it should be the former, but I think you are passing the latter.

> This looks weird. I think you are always resetting the right-most three digits to 0, effectively making the number to 0.

Sorry, never mind. I confused `entry.getMode()` as `layerEntry.getPermissions()`.

Reply:
Judging by the [source of `TarArchiveEntry`](https://commons.apache.org/proper/commons-compress/jacoco/org.apache.commons.compress.archivers.tar/TarArchiveEntry.java.html), it would take `setMode(511)`; it has a couple default modes that show the layout it's expecting.

L217-221:
```
/** Default permissions bits for directories */
public static final int DEFAULT_DIR_MODE = 040755;

/** Default permissions bits for files */
public static final int DEFAULT_FILE_MODE = 0100644;
```

Reply:
Ah, interesting. Thanks for the clarification.